nearcore/src/lib.rs

```diff
@@ -110,20 +120,22 @@ fn open_storage(home_dir: &Path, near_config: &NearConfig) -> anyhow::Result<Nod
         res.with_context(|| format!("unable to open database at {}", opener.path().display()))?;
 
     // Check if the storage is an archive and if it is make sure we are too.
-    // If the store is not marked as archive but we are an archival node that is
-    // fine and we just need to mark the store as archival.
     let hot = storage.get_store(Temperature::Hot);
     let store_is_archive: bool =
         hot.get_ser(DBCol::BlockMisc, near_store::db::IS_ARCHIVE_KEY)?.unwrap_or_default();
-    let client_is_archive = near_config.client_config.archive;
-    anyhow::ensure!(
-        !store_is_archive || client_is_archive,
-        "The node is configured as non-archival but is using database of an archival node."
-    );
-    if !store_is_archive && client_is_archive {
-        let mut update = hot.store_update();
-        update.set_ser(DBCol::BlockMisc, near_store::db::IS_ARCHIVE_KEY, &true)?;
-        update.commit()?;
+    if store_is_archive != near_config.client_config.archive {
+        if store_is_archive {
+            tracing::warn!(
+                target: "neard",
+                "Opening an archival database.  \
+                 Ignoring `archive` client configuration and forcing the node to run in archive mode."
+            );
+            near_config.client_config.archive = true;
+        } else {
+            let mut update = hot.store_update();
```

Comment:
If I have an 'RPC' node that is up to date - and then I change the config and restart -- it will start claiming that it is an archival node - but at the same time, it will not attempt to download the historical data, right?

I'm thinking about a different approach:
* we crash if storage setting != client_config setting
* we add the option to neard to 'change' from archival node to regular node.
* and of course if you start archival node 'from scratch/empty database' it should work.

Reply:
> If I have an 'RPC' node that is up to date - and then I change the config and restart -- it will start claiming that it is an archival node - but at the same time, it will not attempt to download the historical data, right?

Correct.  This is current behaviour as well.

> we crash if storage setting != client_config setting

This is a bit tricky because currently we cannot rely on the storage setting.  IS_ARCHIVE flag was added sneakily without introducing database migration.  This means that it is theoretically normal to open a database without IS_ARCHIVE set but node configured to run in archival mode.

> we add the option to neard to 'change' from archival node to regular node.

This is currently impossible because we don’t generate TrieChanges in archival nodes.

Part of the motivation here is that in the future the `archive` setting will be completely redundant.  Whether node is archival or not will be determined by the node having cold storage configured or not.  This would make the cases quite simple:
- just hot database configured → make sure kind is RPC and run in non-archival mode,
- hot and cold database configured → make sure kinds are HOT and COLD respectively and run in archival mode.

With that we can have a tool which switches between RPC and HOT database kinds in case someone wants to change configuration of their node.

Reply:
@mm-near, friendly ping.